commit 4b1a29a7f5425d32640b34b8a755f34e02f64d0f
Author: Masami Hiramatsu <mhiramat@kernel.org>
Date:   Sat Jan 13 02:56:03 2018 +0900

    error-injection: Support fault injection framework
    
    Support in-kernel fault-injection framework via debugfs.
    This allows you to inject a conditional error to specified
    function using debugfs interfaces.
    
    Here is the result of test script described in
    Documentation/fault-injection/fault-injection.txt
    
      ===========
      # ./test_fail_function.sh
      1+0 records in
      1+0 records out
      1048576 bytes (1.0 MB, 1.0 MiB) copied, 0.0227404 s, 46.1 MB/s
      btrfs-progs v4.4
      See http://btrfs.wiki.kernel.org for more information.
    
      Label:              (null)
      UUID:               bfa96010-12e9-4360-aed0-42eec7af5798
      Node size:          16384
      Sector size:        4096
      Filesystem size:    1001.00MiB
      Block group profiles:
        Data:             single            8.00MiB
        Metadata:         DUP              58.00MiB
        System:           DUP              12.00MiB
      SSD detected:       no
      Incompat features:  extref, skinny-metadata
      Number of devices:  1
      Devices:
         ID        SIZE  PATH
          1  1001.00MiB  /dev/loop2
    
      mount: mount /dev/loop2 on /opt/tmpmnt failed: Cannot allocate memory
      SUCCESS!
      ===========
    
    Signed-off-by: Masami Hiramatsu <mhiramat@kernel.org>
    Reviewed-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Alexei Starovoitov <ast@kernel.org>

diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 2a33efdd1fea..890d4766cef3 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -1551,6 +1551,16 @@ config FAIL_FUTEX
 	help
 	  Provide fault-injection capability for futexes.
 
+config FAIL_FUNCTION
+	bool "Fault-injection capability for functions"
+	depends on FAULT_INJECTION_DEBUG_FS && FUNCTION_ERROR_INJECTION
+	help
+	  Provide function-based fault-injection capability.
+	  This will allow you to override a specific function with a return
+	  with given return value. As a result, function caller will see
+	  an error value and have to handle it. This is useful to test the
+	  error handling in various subsystems.
+
 config FAULT_INJECTION_DEBUG_FS
 	bool "Debugfs entries for fault-injection capabilities"
 	depends on FAULT_INJECTION && SYSFS && DEBUG_FS